### ISSUE TITLE
Better Wayland screenshare support

### ISSUE THREAD

**Issue by MiMillieuh** [Vencord/Vesktop#77]
Hello, I've noticed some weird behaviour with Vencord when screen sharing on Wayland.

But first it works and that's one of the rare client that has Wayland screen share that works.

The issue is that it trigger a lot of portal request to get a screen share and also trigger window and screen at the same time.

As I show in the video below, it's possible to stream but if only one of the steps I do is missed I just have to restart Vencord since it won't trigger screen share any more.

If I succeed I can stop and restart the screen share without any issue (except the one described before).

Here infos about my system : 

```
           .;ldkO0000Okdl;.              mimillie@susemicroos-pro-x 
       .;d00xl:^''''''^:ok00d;.          -------------------------- 
     .d00l'                'o00d.        OS: openSUSE MicroOS x86_64 
   .d0Kd'  Okxol:;,.          :O0d.      Host: 82TL Yoga Slim 7 ProX 14ARH7 
  .OKKKK0kOKKKKKKKKKKOxo:,      lKO.     Kernel: 6.4.6-1-default 
 ,0KKKKKKKKKKKKKKKK0P^,,,^dx:    ;00,    Uptime: 9 hours, 14 mins 
.OKKKKKKKKKKKKKKKKk'.oOPPb.'0k.   cKO.   Packages: 1639 (rpm), 51 (flatpak) 
:KKKKKKKKKKKKKKKKK: kKx..dd lKd   'OK:   Shell: bash 5.2.15 
dKKKKKKKKKKKOx0KKKd ^0KKKO' kKKc   dKd   Resolution: 2560x1440 
dKKKKKKKKKKKK;.;oOKx,..^..;kKKK0.  dKd   DE: GNOME 44.3 
:KKKKKKKKKKKK0o;...^cdxxOK0O/^^'  .0K:   WM: Mutter 
 kKKKKKKKKKKKKKKK0x;,,......,;od  lKk    WM Theme: Adwaita 
 '0KKKKKKKKKKKKKKKKKKKKK00KKOo^  c00'    Theme: Lavanda-Dark [GTK2/3] 
  'kKKKOxddxkOO00000Okxoc;''   .dKk'     Icons: Reversal-purple-dark [GTK2/3] 
    l0Ko.                    .c00l'      Terminal: gnome-terminal 
     'l0Kk:.              .;xK0l'        CPU: AMD Ryzen 7 6800HS Creator Edition (16) @ 3.200GHz 
        'lkK0xl:;,,,,;:ldO0kl'           Memory: 4627MiB / 27731MiB 
            '^:ldxkkkkxdl:^'                                                              
```

https://github.com/Vencord/Vesktop/assets/52078885/edb14f4b-3a8e-472f-93a1-b6659e356704

Thanks !

**Comment by Vendicated** [comment]
you should see #40 for discussion related to wayland streaming. afaik this is still being worked on by electron

**Comment by Vendicated** [comment]
please see if this is any better in the just released v0.2.8

**Comment by MiMillieuh** [comment]
> please see if this is any better in the just released v0.2.8

It's behave the same... :( 

I don't think it is an electron issue on that side... Teams and Webcord uses Electron too and has no issue with Wayland. 

Maybe it's more the implementation used that ask every window and every screen to show them in discord's interface so it always trigger a portal.

**Comment by ToasterUwU** [comment]
discord-screenaudio works perfectly as well. Only one popup which is a combined picker, with a tab for windows and another for screens.

**Comment by Vendicated** [comment]
discord-screenaudio isn't really relevant here because it doesn't use electron

the two pickers one for screens one for monitors is an electron issue that's supposedly fixed, but apparently not in the latest release. the multi step picking where you first pick window in the pipewire picker then inside discord is our issue

**Comment by ToasterUwU** [comment]
Good to know

**Comment by Vendicated** [comment]
I just pushed https://github.com/Vencord/Vesktop/commit/b87bcaefe9fbcc4bbcd0c3ab06e50b331369afd1 which should skip the screen picker inside Discord entirely. Please let me know if this works properly for you!

**Comment by MiMillieuh** [comment]
It still asks for it twice before and also I can do with half a chance a window screen share or a full screen screen share.

I can't choose it depends on what does the last picker do.

the two first picker are useless.


https://github.com/Vencord/Vesktop/assets/52078885/0163f154-a50a-418a-a35f-386850688b52

**Comment by ToasterUwU** [comment]
Armcord had a combined picker working in the newest version yesterday. Think they had to revert it because it would crash for some people, but for me it worked fine.
No extra questions, no endless repicking, no picking screen and window, just one of them.

**Comment by Vendicated** [comment]
yes these are electron issues we need to wait for them to fix. but based on your video it no longer shows the screen picker inside discord, so that's good

**Comment by MiMillieuh** [comment]
> 

It crashes for me :(

**Comment by MiMillieuh** [comment]
> Armcord had a combined picker working in the newest version yesterday. Think they had to revert it because it would crash for some people, but for me it worked fine. No extra questions, no endless repicking, no picking screen and window, just one of them.

It crashes for me :(

**Comment by MiMillieuh** [comment]
> yes these are electron issues we need to wait for them to fix. but based on your video it no longer shows the screen picker inside discord, so that's good

We'll see, I'm just confused that it happens here since it doesn't on teams, skype, zoom, webcord.

these are all electron based applications but they prompt the screen picker correctly.

**Comment by Blisterexe** [comment]
just wanted to butt in, something similar happens for me, it works, but is janky
[Screencast from 2023-08-17 13-14-40.webm](https://github.com/Vencord/Vesktop/assets/128065612/6007748f-717c-4bdc-828d-b1ee5584cf07)
running fedora 38 on wayland with amd gpu.

**Comment by Vendicated** [comment]
this should be considerably improved in vesktop 0.3.1 as it updates electron to 26

**Comment by MiMillieuh** [comment]
> this should be considerably improved in vesktop 0.3.1 as it updates electron to 26

I can't really say due to #125 issue...

**Comment by Blisterexe** [comment]
The screenshare is much improved, but the popup shows twice before the screenshare works, only the second popups settings apply.
If you press "cancel" you have to restart the app to screenshare again.
tested on fedora 38.
[Screencast from 2023-09-25 15-45-40.webm](https://github.com/Vencord/Vesktop/assets/128065612/658e62e8-20f7-4000-b0dd-5e6a3ba55050)

**Comment by MiMillieuh** [comment]
Unfortunately I still have 3 popups like in the original issue on v0.3.2...

**Comment by ToasterUwU** [comment]
i also still get two separate popups, one for window, one for screen. 

Meanwhile Armcord has a combined picker. So its definitely possible to get this into one picker.

**Comment by Vendicated** [comment]
> i also still get two separate popups, one for window, one for screen.
> 
> Meanwhile Armcord has a combined picker. So its definitely possible to get this into one picker.

because we had to downgrade electron again (0.3.2) since it was causing freezes. it's fixed in newer electron (0.3.1) but we cannot use it in this state

**Comment by Vendicated** [comment]
Closing this as completed because there is nothing else we can do. it's entirely fixed in electron 26, so it will be fixed whenever we are able to update

next release, you will also be able to stream in full hd 60fps & with audio on wayland
